lib/Dialect/Arc/Transforms/LowerState.cpp

```diff
@@ -644,6 +596,55 @@ void LowerStatePass::runOnOperation() {
        llvm::make_early_inc_range(getOperation().getOps<HWModuleOp>()))
     if (failed(runOnModule(op)))
       return signalPassFailure();
+
+  // Lower remaining MemoryReadPort ops to MemoryRead ops. This can occur when
+  // the fan-in of a MemoryReadPortOp contains another such operation and is
+  // materialized before the one in the fan-in as the MemoryReadPortOp is not
+  // marked as a fan-in blocking/termination operation in `shouldMaterialize`.
+  // Adding it there can lead to dominance issues which would then have to be
+  // resolved instead.
+  SetVector<DefineOp> arcsToLower;
+  OpBuilder builder(getOperation());
+  getOperation()->walk([&](MemoryReadPortOp memReadOp) {
+    if (auto defOp = memReadOp->getParentOfType<DefineOp>())
+      arcsToLower.insert(defOp);
+
+    builder.setInsertionPoint(memReadOp);
+    Value newRead = builder.create<MemoryReadOp>(
+        memReadOp.getLoc(), memReadOp.getMemory(), memReadOp.getAddress());
+    memReadOp.replaceAllUsesWith(newRead);
+    memReadOp.erase();
+  });
+
+  SymbolTableCollection symbolTable;
+  mlir::SymbolUserMap userMap(symbolTable, getOperation());
+  for (auto defOp : arcsToLower) {
+    auto *terminator = defOp.getBodyBlock().getTerminator();
+    builder.setInsertionPoint(terminator);
+    builder.create<func::ReturnOp>(terminator->getLoc(),
+                                   terminator->getOperands());
+    terminator->erase();
+    builder.setInsertionPoint(defOp);
+    auto funcOp = builder.create<func::FuncOp>(defOp.getLoc(), defOp.getName(),
+                                               defOp.getFunctionType());
+    funcOp->setAttr("llvm.linkage",
+                    LLVM::LinkageAttr::get(builder.getContext(),
+                                           LLVM::linkage::Linkage::Internal));
+    funcOp.getBody().takeBody(defOp.getBody());
+
+    for (auto *user : userMap.getUsers(defOp)) {
+      builder.setInsertionPoint(user);
+      ValueRange results = builder
+                               .create<func::CallOp>(
+                                   user->getLoc(), funcOp,
+                                   cast<CallOpInterface>(user).getArgOperands())
+                               ->getResults();
+      user->replaceAllUsesWith(results);
+      user->erase();
+    }
+
+    defOp.erase();
+  }
```

Comment:
I'm wondering if we can avoid the need for a `FuncOp` in the future if we made `MemoryReadOp` not have a `MemoryRead` side-effect, since LegalizeStateUpdate guarantees that reads do the right thing even if they are misordered according to their side-effects. Sounds ugly though...

Alternatively, we could think about converting *all* arcs into funcs at this point?

Reply:
On the one hand I would love to not have func.func operations at that stage already, on the other hand it would allow us to move read and write operation into the function and thus reducing the number of argument and result values of a call and the number of overall functions because we could merge arcs that are only separated by a memory effect operation.

Changing `MemoryReadOp` to not have a memory read effect is likely not safe though. It's definitely fine before LegalizeStateUpdate, but we also use it after that pass and then, e.g., CSE, could change semantics. Another option could be lowering memories as part of LegalizeStateUpdate, not sure if I like that approach though.

Lowering all functions here already could totally be a way to go. I wasn't really sure what's the best way to deal with the memory side effect added in LowerState, so it would be great if we could discuss this in more detail at some point. My reasoning to not lower all arcs was that `func.call` has arbitrary side-effects while `arc.call` doesn't have any, so it could allow for more canonicalizations/optimizations, but having both at the same time might not be nice indeed.